src/ty_module/generic_arguments.md

```diff
@@ -0,0 +1,126 @@
+definition of that name, and not carried along “within” the type itself).
+
+[nominal]: https://en.wikipedia.org/wiki/Nominal_type_system
+
+
+## The `GenericArgs` type
+
+Given a generic type `MyType<A, B, …>`, we have to store the list of generic arguments for `MyType`.
+
+In rustc this is done using [GenericArgs]. `GenericArgs` is a thin pointer to a slice of [`GenericArg`] representing a list of generic arguments for a generic item. For example, given a `struct HashMap<K, V>` with two type parameters, `K` and `V`, the `GenericArgs` used to represent the type `HashMap<i32, u32>` would be represented by `&'tcx [tcx.types.i32, tcx.types.u32]`.
+
+`GenericArg` is conceptually an `enum` with three variants, one for type arguments, one for const arguments and one for lifetime arguments.
+In practice that is actually represented by [`GenericArgKind`] and [`GenericArg`] is a more space efficient version that has a method to
+turn it into a `GenericArgKind`.
+
+The actual `GenericArg` struct stores the type, lifetime or const as an interned pointer with the discriminant stored in the lower 2 bits.
+Unless you are working with the `GenericArgs` implementation specifically, you should generally not have to deal with `GenericArg` and instead
+make use of the safe [`GenericArgKind`](#genericargkind) abstraction obtainable via the `GenericArg::unpack()` method.
+
+In some cases you may have to construct a `GenericArg`, this can be done via `Ty/Const/Region::into()` or `GenericArgKind::pack`.
+
+```rust,ignore
+// An example of unpacking and packing a generic argument.
+fn deal_with_generic_arg<'tcx>(generic_arg: GenericArg<'tcx>) -> GenericArg<'tcx> {
+    // Unpack a raw `GenericArg` to deal with it safely.
+    let new_generic_arg: GenericArgKind<'tcx> = match generic_arg.unpack() {
+        GenericArgKind::Type(ty) => { /* ... */ }
+        GenericArgKind::Lifetime(lt) => { /* ... */ }
+        GenericArgKind::Const(ct) => { /* ... */ }
+    };
+    // Pack the `GenericArgKind` to store it in a generic args list.
+    new_generic_arg.pack()
+}
+```
+
+[list]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.List.html
+[`GenericArg`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.GenericArg.html
+[`GenericArgKind`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.GenericArgKind.html
+[GenericArgs]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/type.GenericArgs.html
+
+So pulling it all together:
+
+```rust,ignore
+struct MyStruct<T>(T);
+type Foo = MyStruct<u32>
+```
+
+For the `MyStruct<U>` written in the `Foo` type alias, we would represent it in the following way:
+
+- There would be an `AdtDef` (and corresponding `DefId`) for `MyStruct`.
+- There would be a `GenericArgs` containing the list `[GenericArgKind::Type(Ty(u32))]`
+- This is one `TyKind::Adt` containing the `AdtDef` of `MyStruct` with the `GenericArgs` above.
```

Comment:
I don't like the way this last sentence is phrased. Reread it twice, and it does make sense but I think it can be phrased better.

Reply:
oh that should say `there`